### ISSUE TITLE
Feature request: Generate public links for multiple output formats, using multiple renditions.

### ISSUE THREAD

**Issue by jorgenjanssonlee** [robhabraken/content-hub-focal-point-cropping#18]
Hi Rob,

It would be really great if we could define multiple renditions, e.g. jpg_for_web and webp_for_web, that are both used by the script to generate public links in multiple output formats.
I think there are many use cases for this, but one we've come across specifically is a requirement to use webp format for the website, but jpg for a mobile app and/or eDM.

I'll have a go at adding this to the script myself when I have a bit more time to spare in the future, but thought I'd raise it with you as well in case you like the idea.

**Comment by jorgenjanssonlee** [comment]
Ideally, I think there should be a function to request a filetype conversion ad-hoc via the transformations, but that's obviously something I need to raise with Sitecore... :)

**Comment by robhabraken** [comment]
Hi Jorgen,

Thanks for the suggestion. We indeed did also discuss this internally. The reason we didn't choose to go this route, is the sheer amount of public links we would be generating in this case. With a large set of assets (100k in our case), and roughly 10 cropping ratios, doubling the amount of pre-generated links would be quite an impact on storage and traffic in Content Hub. My module kind of forces generation of a lot of public links in Content Hub (of which a lot may be never used) opposed to the default implementation where you would have one or a few public links and generate the rest manually on demand.

Adding multiple rendition sources isn't a big deal in terms of complexity. I already have embedded an automatic switch for vector images. Though it would need a bit more refactoring than only adding generation, because the originalWidth and originalHeight could vary between multiple renditions (the script needs to take that scenarion into account), meaning you basically also need to duplicate all the calculations (lines 270, 272, 273, 277, 278, 313, 314, 316, 321, 322, 326, 327, 341, 342). Additionally, you would also have to maintain an array of scaleFactors. As this is now only one variable used 322 and 327, but this would, when using multiple renditions, also be variable between source images in relation to the preview size used to plot the focal point on. So it can certainly be done, but be careful not to only change the renditionResource in CreatePublicLink (line 196).

Let me think about this for a bit, because it would be most elegant to create another layer of abstraction in the script and separate the calculations from the rendition data retrieval.

**Comment by jorgenjanssonlee** [comment]
Ah yes, thanks for pointing out the scale factor, I think I have an issue with this for a somewhat related change I've done on our end. I wanted to use a single rendition (webp conversion) for all asset media, so I changed the renditionSource and forced Images, Vectors and Gifs through the Vector flow. So they are all using the vector scale factor now. And I think I saw some unexpected crop results and now I understand why. I will do some further testing on this, and some more in-depth reading of the blog post where you explain how the scale factor was worked out and applied.

**Comment by robhabraken** [comment]
No problem. In the case you now describe of pushing all image types through the vector flow, also be aware that for images and vectors you need to retrieve the size properties from different attributes. You can see that in the code too, for images and the default rendition on line 114 I get the width like this:

var originalWidth = mainFile["properties"]["width"].ToObject<int>();

But for vector images, I do this on line 130:

renditions[renditionResource]["properties"]["width"].ToObject<int>();

Additionally, you should take into account that different dimensions also can occur when handling rendition sources that are 300 dpi (mostly with Vector) or 72 dpi. So the image might appear to be equal in size but it is actually larger in pixels, also throwing of the scale factor.

**Comment by jorgenjanssonlee** [comment]
Hi Rob, this is now going way off topic from the initial issue, but I'll keep the conversation going here anyway as it's all related.
The customer has changed their requirements and now want to produce webp public links (based on webp custom rendition) for all input file formats except png and svg. For those, they want to retain the original format (originalDownload rendition). This makes sense to me. But it causes further complications in the script, as the "switch" in rendition source is based on the media flow grouping of various file formats, and I need to split it on a level below. I will have a think about best way to implement this, considering the implications on scale factors. Just thought I'd keep you in the loop of where I'm heading with this. :)

**Comment by robhabraken** [comment]
Hi Jorgen,

- if it is only one public link, uncropped, for webp, you could indeed simply add it based on a separate rendition like the "Original" without apply cropping definition: https://github.com/robhabraken/content-hub-focal-point-cropping/blob/main/Scripts/DAM%20-%20Create%20public%20links%20for%20croppings.cs#L261
- if you need to do all croppings on both webp and the original rendition, you will need to move the scaling variables into arrays for each rendition
- yes, transparancy is retained :-), I already tested this on an earlier version for my vector images

Do you want to finish the addition yourself or would you like to have me chime in? If I create this rendition source array, the code would probably difficult to merge because I may choose a different angle / perspective, and I don't want to set you back to square one.

**Comment by jorgenjanssonlee** [comment]
I would love for you to chime in on this! I will add the png/svg link option myself and share it with you, but the scaling factors are getting hard for me to understand to a level where I'm confident making any changes. I will run this past the developers on my side though, and maybe they will have something to contribute.

**Comment by robhabraken** [comment]
Okay great! I had a first glance a little bit more in detail at how I would approach this, but the script does need even more refactoring than I initially thought. Because the creation of the public link is also affected, because that is based on a certain rendition, but that also means we need to start creating variations in public link names for the same cropping but different renditions, and we also need to pick up on those with the part of the script that determines if it should create or update a public link. I will have to do some other work first though, both community and day job stuff.

**Comment by jorgenjanssonlee** [comment]
Yeah, similar on my side, I will need to put this on ice for a bit while I work through other priorities. But will come back to it after easter.

**Comment by jorgenjanssonlee** [comment]
Hi Rob, feel free to close this one out if you want. What we have in place works well and I don't foresee needing this extra level of complexity for other use cases.

**Comment by robhabraken** [comment]
Thanks Jorgen, I left it open for possible future implementation or use cases, but I didn't run into one yet, and since you now indicate you have found a way around this, and also given the complexity of the change, I will indeed close this for now.
